SourceX/DiabloUI/diabloui.cpp

```diff
@@ -577,27 +599,27 @@ BOOL UiCreatePlayerDescription(_uiheroinfo *info, DWORD mode, char *desc)
 
 void DrawArt(int screenX, int screenY, Art *art, int nFrame, DWORD drawW)
 {
-	if (screenY >= SCREEN_WIDTH || screenX >= SCREEN_WIDTH)
+	if (screenY >= SCREEN_Y + SCREEN_HEIGHT || screenX >= SCREEN_X + SCREEN_WIDTH)
 		return;
 
-	BYTE *__restrict src = (BYTE *)art->data + (art->width * art->height * nFrame);
-	BYTE *__restrict dst = &gpBuffer[screenX + 64 + (screenY + SCREEN_Y) * BUFFER_WIDTH];
-	drawW = drawW ? drawW : art->width;
+	SDL_Rect src_rect = { 0, nFrame * art->h(), art->w(), art->h() };
+	if (drawW && drawW < src_rect.w)
+		src_rect.w = drawW;
+	SDL_Rect dst_rect = { screenX + SCREEN_X, screenY + SCREEN_Y, src_rect.w, src_rect.h };
 
-	const DWORD iSize = std::min(art->height, static_cast<DWORD>(SCREEN_HEIGHT - screenY));
-	const DWORD jSize = std::min(art->width, std::min(drawW, static_cast<DWORD>(SCREEN_WIDTH - screenX)));
+	if (art->surface->format->BitsPerPixel == 8 && art->palette_version != pal_surface_palette_version) {
+#ifdef USE_SDL1
+		if (SDL_SetPalette(art->surface, SDL_LOGPAL, pal_surface->format->palette->colors, 0, 256) != 1)

```

Comment:
We should probably assign the surface it's palette when loading the PCX, each file actually has the correct palette we just didn't need it previously.

Reply:
That's what I tried at first but it didn't work (black screen).

I think it's because of this:

https://github.com/diasurgical/devilutionX/blob/e63657747f4938ce9e8ddfac6036e4bbb4d561a4/SourceX/DiabloUI/diabloui.cpp#L739-L740

Hopefully you can figure it out later.

Reply:
Yeah, a few things have to be rearranged. But this is a really good start, much easier to work with now, thank you.